### PR TITLE
PCHR-3133: Report configuration changes

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview
This PR adds a few fixed requests in [this comment](https://github.com/compucorp/civihr-employee-portal/pull/414#issuecomment-355333041) including:

* The configuration options were aligned to the "report builder" title.
* The configurations list was expanded to 100% so it's aligned with the chart type select and its own save/delete options.
* The configuration options (save, update, and delete buttons) are displayed depending on the context. If there are no configurations saved only "Save as New" is displayed.

In addition this PR fixes these two issues:

* The report would still be visible when switching to the "View Data" tab.
* Field instructions would still be visible when switching report configurations.

### Before
leave reports hr17 9000 1

### After
leave reports hr17 9000

### Before
anim

### After
anim

Only "Save as New" is displayed when there are no configurations saved.

### Technical details

Drupal.behaviors.civihr_employee_portal_reports was refactored so the `attatch` sequence is split into sub methods. This was done for readability's sake:

```js
Drupal.behaviors.civihr_employee_portal_reports = {
  attach: function (context, settings) {
    this.instance = new HRReport();
    this.instance.initAngular();
    this.bindReportConfigurationEvents();
    this.bindTabsEvents();
    this.switchToTabSpecifiedOnTheUrl();
    this.displayConfigurationOptionsIfConfigurationsHaveBeenSaved();
  }
  // ...
}
```

The `displayConfigurationOptionsIfConfigurationsHaveBeenSaved` will show or hide the configuration options depending on the availability of configurations:

```js
displayConfigurationOptionsIfConfigurationsHaveBeenSaved: function () {
  var deleteOption, hasSavedConfigurations, updateOption;

  deleteOption = $('.report-config-delete-btn');
  updateOption = $('.report-config-save-btn');
  hasSavedConfigurations = $('.report-config-select option').length >= 2;

  if (hasSavedConfigurations) {
    deleteOption.fadeIn('fast');
    updateOption.fadeIn('fast');
  } else {
    deleteOption.fadeOut('fast');
    updateOption.fadeOut('fast');
  }
}
```
When a configuration is saved and there was none the configuration options need to be updated. A callback function was added to both `configSaveNew` and `configDelete` to accomplish this:

```js
bindReportConfigurationEvents: function () {
  var self = this;

  $('.report-config-select').bind('change', function (e) {
    self.instance.configGet();
  });
  $('.report-config-save-btn').bind('click', function (e) {
    self.instance.configSave();
  });
  $('.report-config-save-new-btn').bind('click', function (e) {
    self.instance.configSaveNew(function () {
      self.displayConfigurationOptionsIfConfigurationsHaveBeenSaved();
    });
  });
  $('.report-config-delete-btn').bind('click', function (e) {
    self.instance.configDelete(function () {
      self.displayConfigurationOptionsIfConfigurationsHaveBeenSaved();
    });
  });
}
```

### Before
anim

### After
anim

### Technical details
the `displayInstructionsIfDroppableHasNoFields` was created so it displays or hides the instructions depending on the presence of fields in their container:

```js
HRReport.prototype.displayInstructionsIfDroppableHasNoFields = function () {
  $('.pvtAxisContainer .instructions').each(function () {
    var hasFields = $(this).siblings().length;

    if (hasFields) {
      $(this).slideUp();
    } else {
      $(this).slideDown();
    }
  });
};
```

The `displayInstructionsIfDroppableHasNoFields` method is called after dropping a field in a container:

```js
draggableItems.on('sortstop', function (event) {
  droppableContainers.removeClass(highlightClass);
  this.displayInstructionsIfDroppableHasNoFields();
}.bind(this));
```
And after the whole report has been recreated:

```js
HRReport.prototype.updateCustomTemplate = function () {
  var hasReportSectionElement = this.pivotTableContainer.find('.report-section').length;

  if (!hasReportSectionElement) {
    this.createReportSectionElement();
    this.moveReportElements();
    this.appendFilters();
    this.bindFilters();
    this.appendFieldsDraggingInstructions();
    this.displayInstructionsIfDroppableHasNoFields(); // <--- here
    this.bindDragAndDropEventListeners();
  }

  this.updateDropdown();
  this.updateFilterbox();
};
```

### Before
anim

### After
anim

### Technical details
The report pivot tablet element (`#reportPivotTable`) was outside of the tab's pane (first introduced on #414 when moving the report elements). This was moved inside the tab's pane for the Report Builder and an horizontal rule was added to separate the header from the content.

```html
<div class="report-block report-builder tab-pane">
  <div class="chr_search-result__header">
    <div class="chr_search-result__total">
      Report Builder
    </div>
  </div>
  <div id="report-filters" ng-controller="FiltersController as filters">...</div>
  <div id="reportPivotTableConfiguration">...</div>
  <hr />
  <div id="reportPivotTable" class="pvtTable-civi"></div> <!-- here -->
</div>
```

---

- BackstopJS tests are giving false positives so the report page has been tested manually.

### CODE REVIEW

File: civihr_employee_portal/css/custom.css [hunk 1]
```diff
@@ -178,6 +178,10 @@ span.appraisals-employee-legend {
   padding-left: 10px;
 }
 
+#reportPivotTableConfiguration .crm_custom-select {
+  width: 100%;
```
Comment (deb1990) [inline civihr_employee_portal/css/custom.css]: Why is this necessary?

File: civihr_employee_portal/css/custom.css [hunk 2]
```diff
@@ -281,7 +285,7 @@ span.appraisals-employee-legend {
   padding: 8px 20px 8px 15px;
   position: relative;
   white-space: normal;
-  width: inherit;
+  width: 100%;
```
Comment (deb1990) [inline civihr_employee_portal/css/custom.css]: Why is this necessary?

File: civihr_employee_portal/js/reports.js [hunk 3]
```diff
@@ -70,6 +70,22 @@
...
 
+  /**
+   * Hides the report rows and columns instructions if there are fields inside
+   * of their container, otherwise displays the instructions.
+   */
+  HRReport.prototype.displayInstructionsIfDroppableHasNoFields = function () {
+    $('.pvtAxisContainer .instructions').each(function () {
+      var hasFields = $(this).siblings().length;
+
+      if (hasFields) {
```
Comment (deb1990) [inline civihr_employee_portal/js/reports.js]: Please use ternary operator here.

File: civihr_employee_portal/js/reports.js [hunk 4]
```diff
@@ -795,46 +812,90 @@
...
+      this.bindTabsEvents();
+      this.switchToTabSpecifiedOnTheUrl();
+      this.displayConfigurationOptionsIfConfigurationsHaveBeenSaved();
+    },
+    /**
+     * Binds report configuration events for changing, saving, updating and
+     * deleting them.
+     */
+    bindReportConfigurationEvents: function () {
+      var self = this;
```
Comment (deb1990) [inline civihr_employee_portal/js/reports.js]: Dont use `self`, instead use `bind`, `call`, `apply`.

File: civihr_employee_portal/js/reports.js [hunk 5]
```diff
@@ -795,46 +812,90 @@
...
+    },
+    /**
+     * Displays the Report Configuration's save/update/delete options depending
+     * on the availability of configurations. If no configurations are available
+     * only the save option is available and the others are hidden.
+     */
+    displayConfigurationOptionsIfConfigurationsHaveBeenSaved: function () {
+      var deleteOption, hasSavedConfigurations, updateOption;
+
+      deleteOption = $('.report-config-delete-btn');
```
Comment (deb1990) [inline civihr_employee_portal/js/reports.js]:
```
var deleteOption = $('.report-config-delete-btn');
var updateOption = $('.report-config-save-btn');
var hasSavedConfigurations = $('.report-config-select option').length >= 2;
```
This way you dont need to instantiate undefined vars and assign them in the next line. Also this will save you 2 lines.

File: civihr_employee_portal/js/reports.js [hunk 6]
```diff
@@ -795,46 +812,90 @@
...
+    },
+    /**
+     * Automatically switches the current selected tab depending on the tab class
+     * provided in the URL.
+     */
+    switchToTabSpecifiedOnTheUrl: function () {
+      var tabSelector = '.report-tabs a';
+      var hash = window.location.hash;
+
+      hash ? tabSelector += '[data-tab="' + hash.substr(1) + '"]' : tabSelector += ':first';
```
Comment (deb1990) [inline civihr_employee_portal/js/reports.js]: `tabSelector += hash ?  '[data-tab="' + hash.substr(1) + '"]' : ':first';`

File: civihr_employee_portal/css/custom.css (hunk 1 above)
Comment (reneolivo) [reply civihr_employee_portal/css/custom.css]: @bmnepali suggested using `width: inherit` instead of `width: 100%` and that worked well for one of the suggestions, but I didn't notice it didn't work for the fields in the row area:

**Before:**
people reports hr17 9000 2

**After:**
people reports hr17 9000 1

File: civihr_employee_portal/css/custom.css (hunk 2 above)
Comment (reneolivo) [reply civihr_employee_portal/css/custom.css]: @bmnepali suggested using `width: inherit` instead of `width: 100%` and that worked well for one of the suggestions, but I didn't notice it didn't work for the fields in the row area:

**Before:**
people reports hr17 9000 2

**After:**
people reports hr17 9000 1

File: civihr_employee_portal/css/custom.css (hunk 1 above)
Comment (deb1990) [reply civihr_employee_portal/css/custom.css]: In that case its better to add `crm_custom-select--full` to the HTML.
